Support Laravel 9 / PHP 8.1

LukeTowers opened the pull request:
Winter is well on its way to supporting Laravel 9! Thank you for your interest in providing testing for this work, please read the following information on how to get started testing it and report any issues that you encounter!

See https://github.com/wintercms/meta/blob/master/l9-upgrade-notes.md for testing instructions and the upgrade guide.

## Must Have

These issues are crucial for a smooth upgrade process and must be completed before 1.2 can go "stable":

### Internals:

- [x] https://github.com/wintercms/storm/pull/46
- [x] https://github.com/wintercms/winter/issues/578#issuecomment-1165505271
- [x] Revert: Default QueryBuilder methods are defined in the Builder class now preventing the user of `$builder->macro()` to override them in userspace, see https://github.com/laravel/framework/pull/37956#issuecomment-993822397 for details.
- [x] https://github.com/wintercms/storm/pull/64
- [x] https://github.com/wintercms/winter/issues/480
- [x] https://github.com/octobercms/library/commit/29ad91712e5d7ac6e2da70751ea22fe3f2ce9acf (evaluate)
- [x] https://github.com/wintercms/storm/pull/81
- [x] https://github.com/wintercms/winter/issues/503

### v1.1.9 Final changes:
- [x] https://github.com/wintercms/storm/pull/88

### Third Party / Dependencies:

- [x] [barryvdh/laravel-debugbar](https://github.com/barryvdh/laravel-debugbar/tree/master/src/Twig) doesn't support Twig v3.x yet, Winter.DebugBar should replace usage of Twig pieces to add support for v3 or a PR should be submitted upstream.



## Post "stable" release:

The following features were added in Laravel 7-9 and require extra work to "winterize" them for easier consumption / usage in Winter projects. These do not delay "stable" release but should be dealt with soon after:

### New Features:

- [ ] Maintenance mode functionality changed in Laravel 8, see https://github.com/laravel/framework/pull/33560. Explore combining backend maintenance mode setting with the framework maintenance mode, or better document the differences between them and what they're each best used for.
- [ ] Core support for CORS: https://laravel.com/docs/7.x/upgrade#cors-support
- [ ] Review Queue monitoring functionality
- [ ] Review RateLimiter functionality
- [ ] Review Prunable/MassPrunable model traits, especially as it relates to identifying models in the application
- [ ] Review Model Broadcasting
- [ ] Review ImmutableDates 
- [x] Finish implementing `create:model`
- [x] Finish implementing `create:migration`
- [x] Review Laravel commands to identify ones that can be marked as replaced in Winter for aliasing and any that need to be "Winterized".
- [x] Review the difference between Laravel registered default commands and Winter ones and decide which Laravel ones to add / modify for use in Winter.
- [ ] Symfony 5.4 added [support for autocompletion in console commands](https://symfony.com/blog/new-in-symfony-5-4-console-autocompletion), add support to all core commands and document it. Note: Currently waiting on documentation from Symfony
- [ ] Newer Laravel versions don't require manual registration of console commands, investigate how we can implement that for modules and plugins and get rid of our registration code for console commands: https://laravel.com/docs/9.x/artisan#registering-commands
- [ ] Opis/Closure also used in the ExtendableTrait functionality to support serializing classes with dynamic methods (Double check that the usage of the encryption key for serializable closures is being correctly applied)
- [ ] https://github.com/wintercms/winter/pull/45
- [x] https://github.com/wintercms/storm/pull/61
- [x] https://github.com/wintercms/winter/pull/481

### Maintenance:

- [ ] Investigate removing helpers that are now present by default in PHP 8.0
- [ ] https://github.com/wintercms/storm/pull/65
- [ ] Laravel 8.12 introduced support for encryptable casts (https://github.com/laravel/framework/pull/34937 & https://github.com/laravel/framework/commit/1cf55138ac28989a4a8f0f2d6d63784a54dd1061), review and decide how the Encryptable trait should work with this. Technically compatible as of changes in https://github.com/wintercms/storm/pull/26
- [ ] Review the Laravel 6 upgrade (https://github.com/octobercms/october/issues/4381 & https://github.com/octobercms/october/pull/4893) to ensure that nothing was overlooked in the previous upgrade.
- [ ] Review calls to $app->getNamespace() that assume a single application namespace (Usually \App) instead of a collection of modules & plugins that all form the namespace. This is typically used for autoloading files and cleaning up output to remove extra unnecessary paths but those tasks should be handled completely differently in Winter CMS.
- [ ] Laravel v7 introduced changes to getOriginal() in the Model class (HasAttributes trait). The Halcyon model base class does not implement this trait but instead implements similar methods. The Halcyon model base class should be refactored to make use of the HasAttributes trait from Eloquent as much as possible and only provide overrides where necessary.
- [ ] Review YamlProcessor implementation and decide if it needs to be cut down even more.
- [ ] Review Winter\Storm\Foundation\Bootstrap\LoadConfiguration -> Illuminate\Foundation\Bootstrap\LoadConfiguration for conformity with Laravel base class (ties into support for `app.env` configuration variable instead of just `APP_ENV`.
- [ ] Review the helpers that we override from Laravel to make them match up as much as possible (https://github.com/laravel/framework/blob/8b2cdb1dc88af4777ae17dd7279936707f40409e/src/Illuminate/Foundation/helpers.php#L820-L837 vs https://github.com/wintercms/storm/blob/develop/src/Support/helpers.php#L28-L41 (for that one we can possibly go back to the Laravel one but extend e() to turn the parameter into a string first.
- [x] https://github.com/wintercms/storm/pull/72
- [ ] https://github.com/wintercms/winter/pull/521
- [x] Check DebugBar dependency for merging of https://github.com/barryvdh/laravel-debugbar/pull/1318

### Documentation:

- [ ] Document new ArraySource trait
- [ ] Document new tests structure
- [ ] Update 1.2 mail docs to reflect the changes to the configuration files and the introduction of the Driver plugins.
- [ ] Event dispatcher now supports closure event handlers and queued closure handlers, review and document

### API Changes:

- [ ] Revise Cache system (see https://github.com/wintercms/winter/commit/9314a4442e2695b3665e2fa2e0cf1788809e02d8)



## Completed Work:

The following items have all been completed as a part of this Laravel 9 upgrade. Great job to everyone who contributed!

### New Features:
- [x] Add support for `#` as a path symbol that represents the themes directory.
- [x] LazyLoad console commands (https://github.com/laravel/framework/pull/34873 & https://github.com/laravel/framework/pull/34925) (Related: https://github.com/wintercms/storm/pull/60)
- [x] Investigate Anonymous Migrations (https://laravel-news.com/laravel-anonymous-migrations) and alias `artisan migrate` to `winter:up` (fixed in https://github.com/wintercms/storm/commit/b16bb179c4d7f56752b1dd5415fb3d03b6a0581c)
- [x] https://github.com/wintercms/storm/pull/40
- [x] https://github.com/wintercms/winter/pull/404

### Maintenance / Cleanup:
- [x] Reorganize tests out of the project tests folder and into module / plugin specific tests folders.
- [x] Reorganize Console commands into the appropriate modules (https://github.com/wintercms/storm/pull/70, https://github.com/wintercms/winter/pull/471, https://github.com/wintercms/winter/issues/270)
- [x] Fix skipped tests
- [x] Review changes to `artisan serve` (included by the framework now instead of required to existing in the project root)
- [x] https://github.com/wintercms/winter/pull/455
- [x] https://github.com/wintercms/winter/pull/581

### Compatibility Layer:
- [x] Review including a polyfill for the removed `{% spaceless %}` tag in Twig v3.0 (https://github.com/wintercms/winter/pull/456)
- [x] Revise YAML processor system to only engage the processors if the original parsing fails (see https://github.com/wintercms/wn-translate-plugin/commit/a219d7ebf5b2ff6097ce3f7c9f06688b665e19e6 for an example of currently valid YAML that fails if it gets preprocessed).
- [x] ~~`Illuminate\Support\Arr` was moved to `Illuminate\Collections\Arr` (https://github.com/laravel/framework/issues/34292#issuecomment-901657091)~~ (namespace remained the same, file location changed)
- [x] TransportManager renamed to MailManager, check to make sure functionality / public API are unaffected
- [x] ErrorHandler changes to typehint / reference "Throwable" instead of "Exception", check to make sure public API / extensions of this behavior continue to work as expected.
- [x] https://github.com/wintercms/winter/pull/522
- [x] https://github.com/wintercms/storm/pull/69
- [x] https://github.com/wintercms/winter/issues/562
- [x] https://github.com/wintercms/winter/issues/567

### Dependencies:
- [x] https://github.com/wintercms/storm/pull/26
- [x] Switching to Assetic as an external dependency removed the change in FilesystemCache where File::chmod($path) is called, investigate if that actually causes issues for us and we need to implement it in an override or if we're fine.
- [x] Review upgrade to Flysystem v3 (v2 support was added in Winter 1.2 WIP branch, v3 needs to be assessed & tested)
- [x] Revert composer dependencies to stable versions
- [x] Bump minimum version of Laravel to 9.1.0 (see https://github.com/laravel/framework/commit/63ca843643e86fb69efc901051ae079c89a7fd09)
- [x] YAML upgraded from 3 to 5.1 (https://github.com/wintercms/storm/pull/21)
- [x] Remove the embedded implementation of Assetic v1 and utilize [Assetic v2](https://github.com/assetic-php/assetic) instead.

<a href="https://gitpod.io/#https://github.com/wintercms/winter/pull/148"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

tschallacka commented:
Question about the three submodules at https://github.com/wintercms/winter/blob/develop/composer.json#L34 

```json
        "winter/wn-system-module": "dev-develop as 1.1.999",
        "winter/wn-backend-module": "dev-develop as 1.1.999",
        "winter/wn-cms-module": "dev-develop as 1.1.999",
```

They all have a dependency on lLravel 6.0
https://github.com/wintercms/wn-cms-module/blob/develop/composer.json#L28  
https://github.com/wintercms/wn-backend-module/blob/develop/composer.json#L28  
https://github.com/wintercms/wn-system-module/blob/develop/composer.json#L28  

Is it intentional that they are individually dependent on Laravel?   
Or would a winter/storm dependency be better and let the Laravel dependency be decided by Storm?

regardless, can I fork the modules and provide pull requests? As they are marked as read only?

LukeTowers commented:
@tschallacka the modules are locked to the version dependency to stop people from accidentally upgrading them when they don't have the required laravel version. The modules are split manually from the main wintercms/winter repo, so you can fork the main repo and target the 1.2 branch to change their dependencies in a PR and then we can mirror the wip/1.2 branch across when your PR is merged so that you can pull them from composer.

Does that all make sense? It's a somewhat complicated process to be fair.

bennothommo commented:
@LukeTowers we could potentially link it to a Storm branch, and let Storm be the arbiter of which Laravel version is required.

tschallacka commented:
Ah, like laravel does with their system, splitting it up in individual libraries.

jaxwilko commented:
For reference, this PR https://github.com/laravel/framework/pull/38481 removed `Illuminate\Mail\Transport\Transport` which causes:

> PHP Fatal error:  Uncaught Error: Class "Illuminate\Mail\Transport\Transport" not found in .../vendor/winter/storm/src/Mail/Transport/MandrillTransport.php:7

LukeTowers committed:
Replace SwiftMailer with Symfony Mailer

LukeTowers committed:
Merge branch 'develop' into wip/1.2

* develop: (90 commits)
  Define signature for "defineProperties" method
  Add definition of "componentDetails" method
  Update `system` french lang files (#366)
  Update `backend` french lang file (#364)
  Update `cms` french lang file (#365)
  Fix clear color picker function, allow empty value
  Trigger "change" event on color picker when value changes
  Use better cursors for readonly color picker
  Adjust "force" option for plugin:remove
  Require explicit confirmations for CLI deletions
  Add PHP 8.1 option, drop other details field
  Add md_line Twig filter to expose Markdown::parseLine() (#370)
  Don't use the Cache facade in the service provider registration process.
  Fix return type hints for CMS Page and Controller (#362)
  Fixes typo in list of available encodings for importing
  Improved Ukrainian translation (#360)
  Constrain PHP version >= 7.2.9 and 8.0.x (#361)
  Improve Latvian translation (#358)
  Docs UI Fix paths to screenshots of diagrams (#359)
  Coerce Inspector set values to string
  ...

# Conflicts:
#	composer.json
#	modules/backend/composer.json
#	modules/cms/composer.json
#	modules/system/composer.json

LukeTowers committed:
Support Symfony 6.

See https://github.com/wintercms/storm/commit/fcecefda3fd3966310306b5799852c53d6330a64

LukeTowers committed:
Align the CmsException with the Winter ExceptionBase class

LukeTowers committed:
Merge branch 'develop' into wip/1.2

* develop:
  Don't try to access the system parameter database table if no database is in use

LukeTowers committed:
Use the VersionYamlParser when parsing the version.yaml file

LukeTowers committed:
Initial support for Flysystem 2.0

LukeTowers committed:
Support for Symfony 6

LukeTowers committed:
Remove mentions of Rackspace and isolate the storage directory used by tests by default

bennothommo commented on config/app.php:
```diff
@@ -132,7 +134,7 @@
     */
 
-    'trustedProxyHeaders' => Illuminate\Http\Request::HEADER_X_FORWARDED_ALL,
+    'trustedProxyHeaders' => 'HEADER_X_FORWARDED_ALL',
```
Should we still use the constant here? There's no garuntee that the string will remain the same?

LukeTowers replied:
The constant has been completely removed.

bennothommo replied:
Looks like the recommendation is to use the full definition of constants instead: https://github.com/symfony/symfony/blob/4537f85963fba75f9ec2aa9a39c048140260f1f9/src/Symfony/Component/HttpFoundation/Request.php#L50-L51

So I think the value would need to be `'trustedProxyHeaders' => Illuminate\Http\Request::HEADER_X_FORWARDED_FOR | Illuminate\Http\Request::HEADER_X_FORWARDED_HOST | Illuminate\Http\Request::HEADER_X_FORWARDED_PORT | Illuminate\Http\Request::HEADER_X_FORWARDED_PROTO`

LukeTowers replied:
That's what it's mapped to in the middleware in the storm repo, I'm switching to a default of the string constant instead because we've already been burnt by them removing a constant which we have no way to polyfill back in, so all default configs from 1.1.4 to 1.1.8 will be forced to make a configuration file change when upgrading to 1.2 because of it or else their entire site will break because the Symfony / Laravel controlled constant no longer exists.

bennothommo commented on modules/backend/controllers/Files.php:
```diff
@@ -64,33 +65,33 @@ public function thumb($code = null, $width = 100, $height = 100, $mode = 'auto',
-            if (is_null($url) && $disk->exists($path)) {
-                $expires = now()->addSeconds(Config::get('cms.storage.uploads.temporaryUrlTTL', 3600));
-                $url = Cache::remember($pathKey, $expires, function () use ($disk, $path, $expires) {
+        if ($url !== false && is_null($url) && $disk->exists($path)) {
```
This condition looks a bit wonky. There's no need for the `$url !== false` if we're checking that `$url` is `null`.

```suggestion
        if (is_null($url) && $disk->exists($path)) {
```

bennothommo commented on modules/system/classes/PluginBase.php:
```diff
@@ -432,7 +432,13 @@ public function getPluginVersion(): string
 
-        if (!File::isFile($versionFile) || !($versionInfo = Yaml::parse(file_get_contents($versionFile))) || !is_array($versionInfo)) {
+        if (
+            !File::isFile($versionFile) ||
+            !($versionInfo = Yaml::withProcessor(new VersionYamlProcessor, function ($yaml) use ($versionFile) {
+                return $yaml->parseFile($versionFile);
+            })) ||
+            !is_array($versionInfo)
+        ) {
```
I think the PSR guidelines recommend boolean operators being at the start of the line if conditions are multi-line.

```suggestion
            !File::isFile($versionFile)
            || !($versionInfo = Yaml::withProcessor(new VersionYamlProcessor, function ($yaml) use ($versionFile) {
                return $yaml->parseFile($versionFile);
            }))
            || !is_array($versionInfo)
        ) {
```

LukeTowers replied:
Got a link to that? I've always done it the above way so that's probably what's present in the core already. I don't think I've seen it done the other way in the core before.

bennothommo replied:
It's defined in PSR-12 (https://www.php-fig.org/psr/psr-12/) although we've only declared to follow PSR-2 at the moment which doesn't define it.

However, it does say they can be at the beginning or end of a line, but not both:

> Boolean operators between conditions MUST always be at the beginning or at the end of the line, not a mix of both.

Their examples show them at the start of the line, but that's probably the preference of the author.

Personally, I like them at the start because it reads to me like:

- Condition
- OR Condition
- AND Condition

I've also deployed code with this format - eg: [here](https://github.com/wintercms/winter/blob/develop/modules/backend/classes/BackendController.php#L80-L84) and [here](https://github.com/wintercms/winter/blob/develop/modules/backend/widgets/Filter.php#L164-L168). Others have [done so](https://github.com/wintercms/winter/blob/develop/modules/backend/widgets/MediaManager.php#L1414-L1417) as well.

From what I can tell, I think Sam and Alexey used your style mostly.

I don't mind either way - I think we just need to make a ruling, adjust everything to suit and then update the code checks to suit.

LukeTowers replied:
Fair points, I like your reasoning for going with that approach, lets standardize on that and add it to the styling standards.

bennothommo committed:
Merge branch 'develop' into wip/1.2

LukeTowers committed:
Update modules/backend/controllers/Files.php

Co-authored-by: Ben Thomson <git@alfreido.com>

LukeTowers committed:
Remove database.useConfigForTesting

Was redundant, should be using environment config (config/testing/*.php) instead

LukeTowers committed:
Update modules/system/classes/PluginBase.php

Co-authored-by: Ben Thomson <git@alfreido.com>

LukeTowers committed:
Skipped failing tests for now, removed timeout on winter:test command, adjusting testing storage directory

LukeTowers committed:
Merge branch 'wip/1.2' of github.com:wintercms/winter into wip/1.2

* 'wip/1.2' of github.com:wintercms/winter:
  Update modules/system/classes/PluginBase.php

LukeTowers committed:
Merge branch 'develop' into wip/1.2

* develop:
  Backport changes from 1.2 and ignore public directory in default .gitignore
  Update custom_styles language from CSS to LESS (#377)

# Conflicts:
#	config/app.php

LukeTowers committed:
Style fix

LukeTowers committed:
Fix library switcher

bennothommo committed:
Extend abstract Yaml Processor

bennothommo commented:
I had a look at the following:

> Review changes to artisan serve (included by the framework now instead of required to existing in the project root)

Because I use `artisan serve` all the time.

My understanding is that we're already covered here, as `serve` is provided by Laravel, and it looks for an `.env` flie which we create anyway. Is there anything else we would need to do here?

LukeTowers commented:
@bennothommo things to look for are what happens when we get rid of the server.php file and if we need to explicitly register the artisan serve command. Then if that's all fine we need to remove the server.php from the repo and include that in the upgrade guide.

bennothommo committed:
Merge branch 'develop' into wip/1.2

jaxwilko committed:
Merge branch 'develop' into wip/config-file

jaxwilko commented:
There is probably going to be an issue with `str_contains`.

php8.0 spec:
![image](https://user-images.githubusercontent.com/31214002/149437966-58e0d269-221b-4065-a71f-d067a61a28bc.png)

Winter's implmentation:
![image](https://user-images.githubusercontent.com/31214002/149437999-d69b97f7-cea2-4d1b-948e-8faab3de501e.png)

The issue here is in `< php8.0` the `str_contains()` method will allow for an array of strings to be passed as the second option, but upon updating to `> php8.0` the user will get an error along the lines of `TypeError: str_contains(): Argument #2 ($needle) must be of type string, array given`.

Not sure we can do much apart from checking the codebase to make sure we're not using `str_contains()` with arrays and documenting it for users.

`Str::contains($haystack, $needles)` can be used as a replacement.

jaxwilko committed:
Added fix to when loading config values to write in env file

jaxwilko committed:
Updated test case with all strings being quoted

bennothommo commented:
@jaxwilko I think we have to, as a rule, encourage the use of the facades, so we can ensure that they always remain compatible and can account for the discrepancy.

tschallacka commented:
In this case I would suggest adding a special check to the error handler to make the exception message useful for resolving the issue.

Something like, upgrade notice! Replace str_contains() with Str::contains() to use an array for needles.

LukeTowers commented:
@tschallacka That might be a good candidate for integrating the Ignition error handler into the core so that we can write those sorts of custom suggestions using their suggestions/solutions framework (plus the error page is just nicer overall)

LukeTowers committed:
Merge branch 'develop' into wip/1.2

LukeTowers committed:
Update dependencies to match Laravel recommendations

LukeTowers committed:
Improve support PHP 8.1 & align return types with Laravel base classes

See laravel/framework#40471

LukeTowers committed:
app config changes from Laravel 9 skeleton

LukeTowers committed:
Bring the default aliases up to date with Laravel 9

Added:
- Arr: No facade required, also wasn't being registered for Laravel or Winter, using Winter's override
- Str: No facade required, using Winter's override
- Broadcast: Added in Laravel 5.4 (https://github.com/laravel/laravel/commit/52f0196fd3f99113fb838f813564e0c78f133b4d)
- Date: Added in Laravel 8.4.14 (https://github.com/laravel/laravel/commit/89b15441a92aa615de9198c54b5d973aa6b374e5)
- Notification: Added in Laravel 5.3 (https://github.com/laravel/laravel/commit/e6cc60349df90190c5c5597e0980bb7f636a3866)
- RateLimiter: Added in Laravel 8.5.22 (https://github.com/laravel/laravel/commit/c636fd0f678008caf30683833b740fb6f31d9de6)

Removed:
- Redis: Removed in Laravel 8.4.4 due to conflicting with the global Redis class from the default Redis driver (https://github.com/laravel/laravel/commit/612d16600419265566d01a19c852ddb13b5e9f4b)

Some of this could be backported but :shrug:

LukeTowers committed:
Add and comment out missing Js facade

Laravel 8.x added a JS helper to turn PHP data into JS data (see https://github.com/laravel/framework/pull/39389) however I'm not convinced of it's utility over json_encode so this is commented out by default for now.

LukeTowers committed:
Sync config defaults with Laravel 9.x skeleton

Changed from Laravel defaults:
- Array configuration items should have their assigner operators aligned and keys should be sorted alphabetically

Removed:
- database.fetch: Removed from the default in Laravel 5.4 (and originally changed to FETCH_OBJ in 5.3) see https://github.com/laravel/laravel/commit/770c41552f07c75450c72099b8feedbd428888fe & https://github.com/laravel/laravel/pull/3815
- filesystems.cloud: Removed from the default in Laravel 8.4.4 since it's not good practice to use Storage::cloud(), should use explicit disk configurations instead. See https://github.com/laravel/laravel/commit/82213fbf40fc4ec687781d0b93ff60a7de536913
- services.mandrill: No longer officially supported since < Laravel 6.x; can use the SMTP driver instead
- services.sparkport: No longer officially supported in Laravel since < Laravel 6.x or Winter since v1.2 (third party package to reprovide driver is available but config should be manually added when desired.)
- services.stripe: Removed in Laravel v5.8.35, see https://github.com/laravel/laravel/commit/83d2ecc0e9cca7ae6989134dede4a5653a19430b

Explicitly not synced:
- Some calls to env(), will be addressed by future work on the ConfigWriter
- auth.php, will be addressed by future work to bring the Winter auth system more in line with the Laravel one
- cors.php, not implemented by default in Winter at the moment, plugins are available
- filesystems.php: storage.links & storage.disks.public not included because storage:link is not supported by Winter yet and further thought is needed for its inclusion.
- mail.php: markdown - Laravel's markdown mail functionality not verified to work within Winter which has it's own set of email templating logic. May be investigated and included later down the road if desired.
- sanctum.php: Laravel Sanctum is an authentication system for Laravel that is completely untested with Winter. Can be evaluated at a later date for inclusion.

LukeTowers committed:
Cleanup console command return types

LukeTowers committed:
Update testing dependencies to match Laravel 9

bennothommo commented on config/session.php:
```diff
@@ -170,43 +175,44 @@
     */
 
-    'secure' => false,
+    'secure' => env('SESSION_SECURE_COOKIE'),
```
```suggestion
    'secure' => env('SESSION_SECURE_COOKIE', false),
```

bennothommo committed:
Apply correct default for cookie.secure config

LukeTowers committed:
Move CMS console commands into the CMS module

Moved the following:
- \System\Console\ThemeInstall -> \Cms\Console\ThemeInstall
- \System\Console\ThemeRemove -> \Cms\Console\ThemeRemove
- \System\Console\ThemeList -> \Cms\Console\ThemeList
- \System\Console\ThemeUse -> \Cms\Console\ThemeUse
- \System\Console\ThemeSync -> \Cms\Console\ThemeSync

LukeTowers committed:
Merge branch 'wip/1.2' of github.com:wintercms/winter into wip/1.2

LukeTowers committed:
Remove server.php

No longer needed as of https://github.com/laravel/framework/commit/94e846b4db46229a7939d9042d2557123fe45035

LukeTowers committed:
Initial support for command autocompletion

Run `artisan completion --help` for instructions on setting up command autocompletion for your terminal.

LukeTowers committed:
Don't load plugin routes if routes are cached

LukeTowers committed:
Merge branch 'develop' into wip/1.2

# Conflicts:
#	modules/system/ServiceProvider.php

LukeTowers committed:
Code style fix

LukeTowers committed:
Run workflows on wip/1.2

LukeTowers committed:
Don't defer registering module routes

Related to https://github.com/wintercms/storm/commit/290303c740fd7c413e14c3c2092a498fa45de854

LukeTowers committed:
Remove debugging code

LukeTowers committed:
Bump phpcs dependency to support // phpcs: comment syntax

LukeTowers committed:
Only normalize plaintext files when generating FileManifests

LukeTowers committed:
Run default config files through ConfigWriter to minimize changed lines on future uses

LukeTowers committed:
Merge branch 'wip/1.2' into wip/config-file

# Conflicts:
#	modules/system/console/WinterEnv.php

LukeTowers committed:
Bring in line with changes to the storm PR

LukeTowers committed:
Fix botched merge

LukeTowers committed:
Fix winter:env test

LukeTowers committed:
Run the default config through ArrayFile parser

This serves the dual purpose of both testing the output of the ArrayFile parser and normalizing the configs so that any future uses of ArrayFile on them will have easier to understand diffs.

These were generated by running the following in Tinker:

use Winter\Storm\Parse\PHP\ArrayFile;

$configPaths = glob(base_path('config/*'));

foreach ($configPaths as $path) {
  if (is_file($path)) {
  	ArrayFile::open($path)->write();
  }
}

LukeTowers committed:
Revert changes to winter:install, ConfigWriter uses ArrayFile internally

LukeTowers committed:
Initial work on refactoring the Twig usage to use separate environments as required

LukeTowers committed:
Refactor CMS Twig environment.

Credit to @SeriousKen for the original work in https://github.com/octobercms/october/pull/3575

LukeTowers committed:
loadTemplate($name) -> load($name)

Required as of Twig v3

LukeTowers committed:
Fix Twig type hints

bennothommo committed:
Retain spaceless tag in Twig (#456)

Keeps the {% spaceless %} ... {% endspaceless %} tag available in Twig 3.x for Winter CMS backwards compatibility.

While it was deprecated and removed in Twig 3.x, it appears to work still, but may need some more testing.

LukeTowers committed:
Polyfill {% filter %} token removed in Twig v3

LukeTowers committed:
Add AssetCombiner aliases for Snowboard (#459)

LukeTowers committed:
Code style fix and replaced use of {% filter %} in the demo theme

LukeTowers committed:
Add tests for polyfilled {% filter %} Twig tag

LukeTowers committed:
Unskip working tests

LukeTowers committed:
Fix twig tests on Windows

LukeTowers committed:
Fix FilterTest on Windows

LukeTowers committed:
Fix Spaceless test on Windows

LukeTowers committed:
Add more detailed failure message to the ExportModelTest

LukeTowers committed:
Fix ExportModelTest on PHP 8.1

LukeTowers committed:
Merge pull request #455 from wintercms/wip/twig-environments

Refactor Twig registration and usage

mjauvin committed:
Fix support for morphOne/hasOne relations in FormModelSaver (#463)

Fixes #413

LukeTowers committed:
Fix support for | page filter.

Credit to @mjauvin

LukeTowers committed:
Include newly added filters-test in test case to fix tests

LukeTowers committed:
Compile client side translations

LukeTowers committed:
Temporarily disable VersionYamlProcessor

LukeTowers committed:
Temporarily disable extra functionality in the VersionYamlProcessor

LukeTowers committed:
Add asset alias for Snowboard extras css (#464)

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

mjauvin committed:
Fix route order with new system.beforeRoute and system.route events (#465)

With the recent change to module routes being registered in their boot methods the Backend & CMS modules were registering their routes before the System module could. 

Due to the greedy nature of the CMS module's route it must always be registered last.

LukeTowers committed:
Switch to using Assetic v3 instead of embedded Assetic v1 (#470)

mjauvin committed:
fix assetic modules path

LukeTowers committed:
Temporarily avoid testing problematic YAML

LukeTowers committed:
Don't attempt to setup the ThemeLog if the settings table can't be found.

LukeTowers committed:
WIP test adjustments

bennothommo committed:
Restore original YAML processor, merge changes

bennothommo committed:
Re-establish full YAML processor with tweaks.

Should hopefully handle most cases of pre-YAML 4.x version schemas.

Co-Authored-By: Marc Jauvin <marc.jauvin@gmail.com>

bennothommo committed:
Prevent testing direct pushes to wip/1.2 - test the PR instead

bennothommo committed:
Debug branches in action

bennothommo committed:
Further debugging

bennothommo committed:
Let's try this...

bennothommo committed:
Move scaffold commands (#471)

Fixes #270. See https://github.com/wintercms/storm/commit/c4cd4402c0f3e6bfbea08145987482655cbe6583

Moves the Scaffolding commands out of Storm, and into the modules that the scaffold generates code for.

As discussed in the issue above, this makes more sense to keep Storm agnostic, and allows for the commands to be hidden if a particular module is not used in a Winter instance.

LukeTowers committed:
Improved support for autocompletion for the plugin:* commands

Also added System\Console\Traits\HasPluginArgument that can be added to any CLI commands that interact with plugins to provide autocompletion for the plugin argument.

LukeTowers committed:
Improve confirmation behavior for destructive actions

LukeTowers committed:
Move the winter:passwd command into the Backend module

Also add suggestion support for the username argument (last 20 most recently updated records)

LukeTowers committed:
Cleanup scaffolding stubs

LukeTowers committed:
Cleanup console docblocks

LukeTowers committed:
Ensure that mix:watch cleans up after itself

bennothommo committed:
Fix failing tests due to plugin:refresh command change

bennothommo committed:
Prevent mix:install from timing out prematurely

LukeTowers commented on modules/system/classes/PluginBase.php:
```diff
@@ -432,7 +432,13 @@ public function getPluginVersion(): string
+        if (
+            !File::isFile($versionFile)
+            || !($versionInfo = Yaml::withProcessor(new VersionYamlProcessor, function ($yaml) use ($versionFile) {
+                return $yaml->parseFile($versionFile);
```
```suggestion
                return $yaml->parse(file_get_contents($versionFile));
```

LukeTowers committed:
Update modules/system/classes/PluginBase.php

LukeTowers commented on modules/system/classes/UpdateManager.php:
```diff
@@ -890,7 +890,7 @@ public function setNotesOutput($output)
     public function requestServerData($uri, $postData = [])
     {
-        $result = Http::post($this->createServerUrl($uri), function ($http) use ($postData) {
+        $result = \Winter\Storm\Network\Http::post($this->createServerUrl($uri), function ($http) use ($postData) {
```
```suggestion
        $result = Http::post($this->createServerUrl($uri), function ($http) use ($postData) {
```

LukeTowers committed:
Update modules/system/classes/UpdateManager.php

LukeTowers commented on modules/system/classes/VersionManager.php:
```diff
@@ -345,7 +345,9 @@ protected function getFileVersions($code)
         $versionFile = $this->getVersionFile($code);
-        $versionInfo = Yaml::parseFile($versionFile);
+        $versionInfo = Yaml::withProcessor(new VersionYamlProcessor, function ($yaml) use ($versionFile) {
+            return $yaml->parseFile($versionFile);
```
```suggestion
            return $yaml->parse(file_get_contents($versionFile));
```

LukeTowers committed:
Update modules/system/classes/VersionManager.php

bennothommo committed:
Suppress exception when mix:watch is terminated

bennothommo committed:
Merge branch 'wip/1.2' of github.com:wintercms/winter into wip/1.2

bennothommo committed:
Code smell

LukeTowers committed:
Merge branch 'develop' into wip/1.2

LukeTowers committed:
Add missing import

LukeTowers committed:
Fix missing import

LukeTowers committed:
More fixes to the mailer

bennothommo committed:
Additional test cases and fixes for Version Yaml parser

bennothommo committed:
Fix unit tests

LukeTowers committed:
Merge branch 'develop' into wip/1.2

# Conflicts:
#	modules/cms/classes/Controller.php

LukeTowers committed:
Merge branch 'develop' into wip/1.2

LukeTowers committed:
Revert "Run default config files through ConfigWriter to minimize changed lines on future uses"

This reverts commit 897263e2813596e116ed0b1350e1a9034a646cb0.

LukeTowers committed:
Merge branch 'wip/1.2' into wip/config-write-replacement

jaxwilko committed:
Merge branch 'wip/1.2' into wip/config-write-replacement

jaxwilko committed:
Merge branch 'wip/config-file' into wip/config-write-replacement

LukeTowers committed:
Merge branch 'wip/config-write-replacement' of github.com:wintercms/winter into wip/config-write-replacement

mjauvin committed:
Use env() settings by default in config/cms.php (#484)

LukeTowers committed:
Initial work on re-aligning the scaffolding commands with Laravel

LukeTowers committed:
Fix issue with refactored Symfony component

LukeTowers committed:
Remove support for varcharmax

This was originally added in https://github.com/wintercms/winter/commit/bd9f88587f5adad66866294d22a598aa195cff4b (and further refined in https://github.com/wintercms/winter/commit/160ae441ff29dfde81e7efd42ac3f1e26f5c7985) as a workaround for issues in MySQL < 5.7 and MariaDB < 10.2 related to the default string length when using the recommended utf8mb4 database encoding.

As of Laravel 8 the minimum supported versions of MySQL and MariaDB are not affected by this issue so we're now removing this configuration / workaround in Winter v1.2.

LukeTowers committed:
Fix scope overreach of migration from 2016.

Backend migrations should only touch the backend tables, same for the system module. This was added for build 345 in 2016 in the wrong migrations. See https://github.com/wintercms/winter/commit/d7cfe61e67bb1e502354442c633e74a26da4b113

Flynsarmy committed:
Strip double / prefix when uploading to Media root directory

Flynsarmy committed:
Make getMediaPath and getStorageDisk methods public

Flynsarmy committed:
Make getMediaPath public

LukeTowers committed:
Merge pull request #491 from Flynsarmy/uploadFilepath

Strip double / prefix when uploading to Media root directory

LukeTowers committed:
Added missing return statement

LukeTowers committed:
Revert "Run the default config through ArrayFile parser"

This reverts commit c3bdfae664c1cdde2ea1403027fc38b0822a0cf5.

LukeTowers committed:
Merge branch 'wip/1.2' into wip/config-write-replacement

LukeTowers committed:
Tweaks to be more friendly with the ArrayFile parser

LukeTowers committed:
Merge branch 'wip/config-write-replacement' into wip/config-file

# Conflicts:
#	config/app.php
#	config/auth.php
#	config/cms.php
#	config/database.php

LukeTowers committed:
Style fix

LukeTowers committed:
Run all config files through ArrayFile parser

This will minimize changes when ConfigWriter is used to set config values through PHP.

LukeTowers committed:
Support new mail config file structure

LukeTowers committed:
Merge pull request #404 from wintercms/wip/config-file

[Enhancement] Refactor winter:env, winter:install to use new EnvFile & ArrayFile parsers

mjauvin committed:
Fix isActive for CMS page with "/" as URL (#479)

Co-authored-by: Luke Towers <github@luketowers.ca>
Co-authored-by: Ben Thomson <git@alfreido.com>

mjauvin committed:
Sync MailSettings model init with L9 mail config changes (#500)

damsfx committed:
Multiple file attachments in theme customization (#495)

This PR add a `multiple` property to fields in [theme customization](https://wintercms.com/docs/themes/development#customization).  
If `multiple: true`, the field is added to `attachMany` and allow multiple files selection, in all other cases the current behaviour is applied.

LukeTowers committed:
Refresh datasource path cache before syncing paths

Fixes an issue where if the underlying data behind a datasource changes through manual intervention (either in the database or the filesystem) before running theme:sync it wasn't being detected by the theme:sync command.

mjauvin committed:
Prevent error if accessing undefined $resizer variable (#502)

LukeTowers committed:
Fixed ImageResizer support for "system" images (plugin, theme, module assets)

Also improved the developer experience by making it easier to see what the actual issue is if an error is encountered during the resizer initialization process.

LukeTowers committed:
Merge branch 'wip/1.2' into wip/1.2-improve-scaffolding

LukeTowers committed:
Simplify diff

LukeTowers committed:
Adjusted interfaces and made create:plugin include translations

LukeTowers committed:
Add localization support to create:controller

LukeTowers committed:
Localized create:model scaffolding

LukeTowers committed:
Simplify prepareVars usage, fix create:theme usage

LukeTowers committed:
Merge develop into 1.2

GrCOTE7 commented:
Hi, all !

I try to apply the documentation about Relationships with this version (Part ManyToMany).

Under the version with Laravel 8, i completly succeed that:

(
Note I have just 2 main tables:
- **grcote7_marriage_guests** (With has only a 1 to 1 relation with **users**) and 4 guests
- **grcote7_marriage_groups** (3 goups: AMI - FAMILLE - TRAVAIL)

and a pivot table named: **grcote7_marriage_group_guest**... as described in the documentation
)

and so, ll is perfect with Laravel 8 :

http://prntscr.com/VXZ1Rs4fvC6h

But, with Laravel 9  (Exactly the same code): 

http://prntscr.com/xf0X0wq6306m

Note:
All the data of this example can easily be reproduced: see the **README.md** of : 

https://github.com/GrCOTE7/prepa

*Hope this can help you to... Help me ! ;-)*

Thanks.

mjauvin commented:
@GrCOTE7 What do you mean Laravel 8 ? The previous version for winter was Laravel 6.

GrCOTE7 commented:
Hi, mjauvin...

You certainly are right...

I mean the version which has got this requirement : 

http://prntscr.com/qgLBiNhBbOr8

And I 'd imagined 'More than 6' and perhaps quick concluded '8' ;-) !

;-)

mjauvin committed:
fix null coalescent operator (#518)

LukeTowers committed:
Merge branch 'wip/1.2' into wip/1.2-improve-scaffolding

LukeTowers committed:
Moved GeneratorCommand back to Scaffold

LukeTowers committed:
Plugins should be generated the same no matter what OS is being used to generate them

LukeTowers committed:
Stub out create:migration

LukeTowers committed:
Fix create:controller command

LukeTowers committed:
Improve Settings scaffolding

LukeTowers committed:
Merge pull request #486 from wintercms/wip/1.2-improve-scaffolding

Scaffolding Improvements

Still a few todo items but I've added them to the main comment on the Laravel 9 PR to be dealt with later so that we can start using the work that exists on this so far.

mjauvin committed:
Update MailSettings for new Laravel mail config structure. (#522)

Requires wintercms/storm#79

Removed the following drivers configs not native in Laravel:
- Mailgun
- Mailchimp (Mandrill)
- Postmark
- SparkPost
- Amazon SES

Credit to @mjauvin

LukeTowers committed:
Minor tweaks; moved SES fields to the Winter.DriverAWS plugin

LukeTowers committed:
Merge pull request #526 from wintercms/wip/1.2-fix-backend-mail-settings

Update MailSettings for new Laravel mail config structure. (#522)

mjauvin committed:
Add "send test message" button on MailSettings page (#528)

Implements #527. Credit to @mjauvin

mjauvin committed:
change confirmation prompt (#529)

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Make "mix:update" command compatible with 1.2

bennothommo committed:
Merge branch 'develop' into wip/1.2

WebVPF committed:
Allow for nested partial files in components in the recommended extension file type mapping (#453)

Credit to @WebVPF

mjauvin committed:
Only show "Send test message" button on Mail Settings page (#537)

bennothommo committed:
Merge branch 'develop' into wip/1.2

LukeTowers committed:
Add support for .php as a backend partial extension

LukeTowers committed:
Use .php for the backend view extension instead of .htm

LukeTowers committed:
Fixes for Windows & linter for ViewMaker trait

bennothommo committed:
Implement code style for partials.

Adjust all partials to meet new code style.

bennothommo committed:
Fix linting issue with fileupload partial

LukeTowers committed:
Fix windows tests & update references to backend view files

mjauvin committed:
hide smtp user/password fields if smtp authorization is false (#544)

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo commented:
@LukeTowers 

> https://github.com/octobercms/library/commit/29ad91712e5d7ac6e2da70751ea22fe3f2ce9acf (evaluate)

FYI, I've made that same change in the code analysis branch. Symfony's `UploadedFile` class no longer allows the size to be passed through. Symfony 4 had a BC layer in place, but 5 dropped it.

https://github.com/symfony/symfony/pull/25471

bennothommo committed:
Compile all assets

bennothommo committed:
Upgrade Font Awesome 4 to 6

bennothommo committed:
Fix shimmed icons, remove some CSS variables, recompile

bennothommo committed:
Remove extra license file

LukeTowers committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Merge branch 'wip/1.2' into wip/1.2-font-awesome-6

bennothommo committed:
Remove explicit links to icon LESS file.

We will load icons independently in another file so it can be cached independently.

bennothommo committed:
Compile icon styling into separate CSS file

bennothommo committed:
Load separate icon file in Backend pages

bennothommo committed:
Move icons into root UI folder.

Also separated the global base styles for icons with the actual icon list, so we can limit the icons loaded (ie. for the standard exception / error views)

bennothommo committed:
Fix font location

bennothommo committed:
Load a subset of icons for the exception / message views

bennothommo committed:
Add docs for icons

LukeTowers committed:
Merge pull request #551 from wintercms/wip/1.2-font-awesome-6

Upgrade Font Awesome 4 to 6

LukeTowers committed:
Fix icon usage for ReportContainers, AssetList, & Table widgets

bennothommo committed:
Fix prefixed shims overriding entire element with FA font

bennothommo committed:
View Maker guessed paths can be null

bennothommo committed:
Optimise shimmed icon CSS

bennothommo committed:
Revert "Optimise shimmed icon CSS"

This reverts commit f8be6d7e85bf0bbadcb19294d7309843f6a0eb1f.

bennothommo committed:
Manually optimise shimmed icon CSS

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Include brand icons, small optimisation

bennothommo committed:
Drop "oc-" prefix for icons introduced in Font Awesome 6.

We're only keeping that prefix for backwards compatibility, so no need for the prefix on new icons. Should bring the filesize down a fair chunk.

bennothommo committed:
Remove Workshop theme from repo.

Accidentally committed in last commit

jaxwilko committed:
Respect app.asset_url in AssetMaker (#552)

LukeTowers committed:
Force resized images to be regenerated when the source image changes (#178)

This should fix https://github.com/wintercms/winter/issues/177, but it will cause all existing installations to resize all of their already resized images. 

Another alternative could have been to include the hash of the source image; but it would have been at the expense of more work every time this function is called, which could potentially include generating hashes of multiple extremely large (10+ MB) images on every page load if the page includes such images that need to be resized since the filter needs to identify the configuration in order to identify the path to the resized image to check if it needs to return a resizer/ URL or a /resized one.

Further improvements could potentially be made in the isResized()

bennothommo committed:
Remove blog plugin from repo

LukeTowers committed:
Normalize versions pulled from the version.yaml

Fixes an issue where a plugin switches from no v prefix in version identifiers to having a v prefix and then gets replaced.

jaxwilko committed:
Small refactor to workflow utilities (#555)

Refactored workflow utilities to use STDOUT & STDERR

LukeTowers committed:
Added TailwindCSS-based scaffold to the create:theme command

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Merge branch 'develop' into wip/1.2

LukeTowers committed:
Improve error handling of invalid values provided to the DatePicker FormWidget

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Allow decompilation of local assets that match local URL

https://github.com/wintercms/winter/pull/552 introduced a change where all backend assets are converted to URLs to provide support for the `app.asset_url` configuration variable. This change allows asset decompilation to still work by converting a (local) URL back to a relative path. If a URL doesn't match the root URL, then it is passed through as is with no decompilation taking place.

Fixes https://github.com/wintercms/winter/issues/562.

LukeTowers committed:
Revamp Plugin Manager (#501)

Reworked the internal disabled status checking and monitoring system for plugins.
Added better support for plugin disabled status management.
Added improvements to loading (should aid in improving app bootup speed)

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Fix code smell

LukeTowers committed:
Improve reliability of cache:clear

This improves the reliability of the cache:clear command by not requiring the custom winter cache directories to exist when running cache:clear.

Also removes the clearing of disabled.json which is no longer used as of https://github.com/wintercms/winter/commit/1bbfb67d8c2d243bf25a79cc5c5ed635360d5e77

mjauvin committed:
fix typo

jaxwilko committed:
Moved plugin replacement namespace aliasing into register replacement method (#580)

Required to ensure that plugins that are marked as invalid to replace others will not have their namespaces aliases.

jaxwilko committed:
[FIX] Fix plugin flags not loading from cache correctly (#582)

Fix issue where flags would not be set correctly from cache

jaxwilko committed:
Split tests into relevant module folders

This splits the testing suite into the separate modules as appropriate in order to improve the reliability of the testing suite as a whole and make it easier for developers to have an up to date testing suite from the core to build off of. Additionally the tests are now namespaced and some minor improvements to the PluginManager were made.

Now the PluginManager will internally treat plugin identifiers as lower case strings, only transforming them to their normalized versions when requested by methods like getPlugins() & getAllPlugins(). The idea behind this is that it provides a much simpler way to internally handle checking, especially for plugin replacement where casing could cause issues.

Replaces #576. Fixes #575.

jaxwilko committed:
Added replaced plugins to the normalize map to ensure classloader namespace aliasing detects replacements (#585)

mjauvin committed:
Add additional testMakePartial cases to ViewMaker unit tests (#586)

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Allow a string selector for the form in a request

LukeTowers committed:
Switch back to using Laravel CacheServiceProvider

Partially reverts part of https://github.com/wintercms/winter/commit/1bbfb67d8c2d243bf25a79cc5c5ed635360d5e77 in favour of https://github.com/wintercms/storm/pull/86/files

bennothommo committed:
Fix site relative partial paths failing in 1.2 (#587)

Matches the behavior of the ViewMaker in the 1.1 branch more closely by checking if the partial path is a local reference first, returning that path if it is. Only if this check fails will it then run through the view paths.

Fixes https://github.com/wintercms/winter/issues/571

Also fixes the ViewMaker tests by emulating the same behaviour as a web request and having the current directory be the base path of the site, not the directory of the current module or plugin being tested.

LukeTowers committed:
Bump minimum Laravel version to 9.1

See https://github.com/laravel/framework/commit/63ca843643e86fb69efc901051ae079c89a7fd09

bennothommo committed:
Merge remote-tracking branch 'origin/develop' into wip/1.2

jaxwilko committed:
Added fix to ensure correct normalization and return (#588)

Fixes https://github.com/wintercms/winter/issues/569

jaxwilko committed:
Use the ImageResizer for resizing and cropping in the MediaManager (#577)

This implements support for the ImageResizer within the media manager for thumbnail resizing & image cropping. This simplifies and centralizes the image resizing code within the media manager and allows for better support on readonly filesystems by relying on the ImageResizer class to process image modification requests.

LukeTowers commented:
@bennothommo can you resolve the conflict with the ViewMaker test?

bennothommo committed:
Merge branch 'develop' into wip/1.2

bennothommo committed:
Code analysis tweaks (#589)

Implements fixes to some breaking changes introduced in https://github.com/wintercms/storm/pull/64.

bennothommo committed:
Fix constructor signature in Meta class

bennothommo committed:
Fix constructor signature in User Throttle model

jaxwilko committed:
Added defaults to brand settings config calls (#595)

RomainMazB committed:
Fix return type in placeholderFunction (#594)

jaxwilko committed:
Added lazy resizing support for thumbnail view in media manager

LukeTowers committed:
Code review

Removed a bunch of thumbnail generation related methods that are no longer required now that the ImageResizer class handles it all.

LukeTowers committed:
Code style fixes

LukeTowers committed:
Merge pull request #596 from wintercms/wip/1.2-fix-mediamanger-thumbnail-view

[WIP] 1.2 - Fix MediaManger Thumbnail View

LukeTowers committed:
Cleanup unnecessary imports

LukeTowers committed:
Remove plugin from pluginFlags property if it has no active flags

Fixes an issue that occurs when removing the last flag from a plugin doesn't re-enable it.

LukeTowers committed:
Default to public visibility for the local disk

See https://github.com/wintercms/winter/issues/503

LukeTowers committed:
Add aliases for base test classes for compatibility

LukeTowers committed:
Add create:job scaffolding command

bennothommo committed:
Add unit test for #578

bennothommo committed:
Actually finish the test

LukeTowers committed:
Fix issues with accessing the expected context variables inside of Twig macros.

This fixes #578 by adding the ability to pass the CMS Controller instance to the CMS Twig Extension removing the reliance on context variables as well as making the expected "global" twig variables inside of the CMS Twig environment actually global within that environment.

Replaces #598 & #593.

Credit to @RomainMazB for the initial implementation.

WebVPF committed:
Improved Russian translation (#605)

LukeTowers committed:
Merge branch 'develop' into wip/1.2
